### PR TITLE
Sd 17

### PR DESCRIPTION
**Pull request by preetamnpr**

### CODE REVIEW

**Comment by jkosternl** [comment]
Hi @preetamnpr, I just checked what SD-17 actually is, and I see: "OVS - write the Dev Portal implementation guide"
I now wonder if that is correct. Can you please rename this PR so it matches the Jira issue and also has a meaningful name? That helps me to efficiently review it and secondly, while merging, your commit message is than automatically also correct. Thanks!

**Comment by jkosternl** [comment]
> I realized it later that the SD-17 was for the developer portal rather than the OVS checks enhancement.

No problem, please just update the title of this PR, so it matches reality. If you squash merge it, the used commit messages don't care!

File: ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/OvsChecks.java [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+      body -> {
+        Optional<Map.Entry<OvsFilterParameter, String>> limitParam =
+          sspSupplier.get().getMap().entrySet().stream()
+            .filter(e -> e.getKey().equals(OvsFilterParameter.LIMIT))
+            .findFirst();
+
+        if (limitParam.isPresent()) {
+          int expectedLimit = Integer.parseInt(limitParam.get().getValue().trim());
+          if (body.size() > expectedLimit) {
+            return Set.of("The number of events exceeds the limit parameter: " + expectedLimit);
```
Comment (palatsangeetha) [inline ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/OvsChecks.java:153]: I think the message can be 'schedules' instead of 'events'

File: core/src/main/java/org/dcsa/conformance/core/party/CustomJsonPointer.java [hunk 2]
```diff
@@ -13,7 +13,8 @@ private CustomJsonPointer() {
     // Private constructor to prevent instantiation
   }
 
-  public static List<JsonNode> findMatchingNodes(JsonNode rootNode, String pathExpression, BiPredicate<JsonNode,String> condition, String paramValue) {
+  public static List<JsonNode> findMatchingNodes(JsonNode rootNode, String pathExpression,
+                                                 BiPredicate<JsonNode,String> condition, String paramValue) {
```
Comment (jkosternl) [inline core/src/main/java/org/dcsa/conformance/core/party/CustomJsonPointer.java:17]: Which formatting are you using, because when I format it, it formats differently. These differences should be avoided.

File: ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/OvsChecks.java [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+            (eventDate, expectedEndDate) ->
+              !eventDate.isBefore(expectedEndDate)))); // Check if eventDate is after or equal to endDate
+
+    checks.add(JsonAttribute.customValidator(
+      "Validate transportCallReference is unique across each array node",
+      OvsChecks::validateUniqueTransportCallReference
+    ));
+
+    checks.add(JsonAttribute.customValidator(
+      "Validate limit parameter is met",
```
Comment (jkosternl) [inline ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/OvsChecks.java:143]: I have no clue what this sentence means for the end user. Can you rephrase it, to make it clear? E.g.: limit parameter is wrong? Limit parameter is higher than allowed? I don't know.

File: ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/OvsChecks.java [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    Supplier<SuppliedScenarioParameters> sspSupplier,
+    OvsFilterParameter dateParameter,
+    BiPredicate<LocalDate, LocalDate> dateComparison) {
+
+    Optional<Map.Entry<OvsFilterParameter, String>> dateParam =
+      sspSupplier.get().getMap().entrySet().stream()
+        .filter(e -> e.getKey().equals(dateParameter))
+        .findFirst();
+
+    if (dateParam.isPresent()) {
```
Comment (jkosternl) [inline ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/OvsChecks.java:220]: Perhaps you can return here when the dateParam is not present. That avoids many indented lines and mental effort. 
See for background info: https://medium.com/swlh/return-early-pattern-3d18a41bba8

File: ovs/src/main/java/org/dcsa/conformance/standards/ovs/action/SupplyScenarioParametersAction.java [hunk 5]
```diff
@@ -3,34 +3,26 @@
...
-                    .collect(Collectors.joining(", "))));
-    this.ovsFilterParameters = new LinkedHashSet<>(Arrays.asList(ovsFilterParameters));
+    String publisherPartyName, Map<OvsFilterParameter, String> parameters) {
+    super(publisherPartyName,null, null,
+      "SupplyScenarioParameters(%s)"
+        .formatted(
+          parameters.entrySet().stream()
+            .map(entry -> entry.getKey().getQueryParamName() + "=" + entry.getValue())
+            .collect(Collectors.joining(", "))), -1 );
+    this.ovsFilterParameterMap =  parameters;
```
[the remark this answers is not in the document]
Comment (preetamnpr) [reply ovs/src/main/java/org/dcsa/conformance/standards/ovs/action/SupplyScenarioParametersAction.java:25]: The below format settings are being in the Intellij ? what formatting settings are you using ? we need to be uniform in order to have the uniform source code format . 
[*]
charset = utf-8
end_of_line = lf
indent_style = space
indent_size = 2
trim_trailing_whitespace = true
insert_final_newline = true

File: ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/QueryParameterSchemaCheck.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static org.dcsa.conformance.core.toolkit.JsonToolkit.OBJECT_MAPPER;
+
+public class QueryParameterSchemaCheck extends ActionCheck {
+
+  private final JsonNode parametersSchema; // Store the "parameters" array from OpenAPI spec
+  private final JsonSchemaFactory factory = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V7);
```
[the remark this answers is not in the document]
Comment (preetamnpr) [reply ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/QueryParameterSchemaCheck.java:26]: I am trying to load the schemas for the query parameters of GET endpoint in order to validate them. These schemas are present in the Open api spec.

File: core/src/main/java/org/dcsa/conformance/core/party/CustomJsonPointer.java (hunk 2 above)
Comment (preetamnpr) [reply core/src/main/java/org/dcsa/conformance/core/party/CustomJsonPointer.java:17]: [*]
charset = utf-8
end_of_line = lf
indent_style = space
indent_size = 2
trim_trailing_whitespace = true
insert_final_newline = true

File: ovs/src/main/java/org/dcsa/conformance/standards/ovs/action/SupplyScenarioParametersAction.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkosternl) [reply ovs/src/main/java/org/dcsa/conformance/standards/ovs/action/SupplyScenarioParametersAction.java:25]: And what if you format the changed lines? Does it change? My point is that it just looks to be not formatted a bit, e.g. at line 15. 

Those lines are the lines as defined in `.editorconfig`. If you did not configure anything manually, and you are using `google-java-format` and enabled it in your project, it should behave the same as my IDE. Can you check these are fine?

[the remark this answers is not in the document]
Comment (preetamnpr) [reply ovs/src/main/java/org/dcsa/conformance/standards/ovs/action/SupplyScenarioParametersAction.java:25]: image
here is the snapshot from my ide.

File: ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/OvsChecks.java (hunk 4 above)
Comment (preetamnpr) [reply ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/OvsChecks.java:220]: The logic is if the dateParams is provided in the query parameters then the check must be performed otherwise nothing must be performed. why should i waste the runtime cycle to check if dateParam is not present for every scenarios and return empty set.   I find this logic must only be executed when the dateParam is present.

File: ovs/src/main/java/org/dcsa/conformance/standards/ovs/party/OvsPublisher.java [hunk 7]
```diff
@@ -95,16 +88,72 @@ private void supplyScenarioParameters(JsonNode actionPrompt) {
   @Override
   public ConformanceResponse handleRequest(ConformanceRequest request) {
     log.info("OvsPublisher.handleRequest(%s)".formatted(request));
+    Map<String, Collection<String>> headers = new HashMap<>(Map.of(API_VERSION, List.of(apiVersion)));
```
[the remark this answers is not in the document]
Comment (preetamnpr) [reply ovs/src/main/java/org/dcsa/conformance/standards/ovs/party/OvsPublisher.java:91]: I agree, during coding I was thinking what variables are required hence defined it before hand. I will move this one before its been used.

File: ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/OvsChecks.java (hunk 3 above)
Comment (preetamnpr) [reply ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/OvsChecks.java:143]: Ok.. will elaborate it more.

File: ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/QueryParameterSchemaCheck.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jkosternl) [reply ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/QueryParameterSchemaCheck.java:26]: If tried to investigate this myself on how these things should be validated, but I didn't get a better result and we should leave it like this. I'll resolve this thread.

**Review by jkosternl — APPROVED** [review]
Looking good, from the Java perspective. Thanks Preetam!
Please wait for merging on the others (different perspectives).

File: ovs/src/main/java/org/dcsa/conformance/standards/ovs/party/OvsPublisher.java [hunk 8]
```diff
@@ -96,15 +88,78 @@ private void supplyScenarioParameters(JsonNode actionPrompt) {
...
         JsonToolkit.templateFileToJsonNode(
             "/standards/ovs/messages/ovs-%s-response.json"
                 .formatted(apiVersion.toLowerCase().replaceAll("[.-]", "")),
             Map.ofEntries());
 
-    return request.createResponse(
-        200,
-        Map.of(API_VERSION, List.of(apiVersion)),
-        new ConformanceMessageBody(jsonResponseBody));
+    ArrayNode filteredArray = OBJECT_MAPPER.createArrayNode();
```
Comment (palatsangeetha) [inline ovs/src/main/java/org/dcsa/conformance/standards/ovs/party/OvsPublisher.java:100]: Is this for dynamically generating the response from the static file according to the filter params?

Comment (preetamnpr) [reply ovs/src/main/java/org/dcsa/conformance/standards/ovs/party/OvsPublisher.java:100]: Yes, indeed.. The publisher is created such a way that it exhibits real behavior.

File: ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/OvsChecks.java [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+            body ->
+                validateParameter(
+                    body,
+                    sspSupplier,
+                    OvsFilterParameter.VESSEL_IMO_NUMBER,
+                    "*/vesselSchedules/*/vesselIMONumber")));
+
+    checks.add(
+        JsonAttribute.customValidator(
+            "Validate vesselName exists and"
```
[the remark this answers is not in the document]
Comment (preetamnpr) [reply ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/OvsChecks.java:66]: The checks are triggered only when the query parameter exists in the request and the json path for searched attribute exists in the response then the value is checked to make sure that publisher indeed provided the right data, otherwise the checks are passed. I hope this answers your question, else connect with me.

[the remark this answers is not in the document]
Comment (palatsangeetha) [reply ovs/src/main/java/org/dcsa/conformance/standards/ovs/checks/OvsChecks.java:66]: Thanks Preetam for clarifying it.
